src/node_crypto.cc

```diff
@@ -2010,6 +2025,91 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
 #endif  // OPENSSL_NPN_NEGOTIATED
 
+#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
+typedef struct tlsextalpnctx_st {
+  unsigned char* data;
+  unsigned short len;
+} tlsextalpnctx;
+
+template <class Base>
+int SSLWrap<Base>::SelectALPNCallback(SSL* s,
+                                      const unsigned char** out,
+                                      unsigned char* outlen,
+                                      const unsigned char* in,
+                                      unsigned int inlen,
+                                      void* arg) {
+  Base* w = static_cast<Base*>(SSL_get_app_data(s));
+  Environment* env = w->env();
+  HandleScope handle_scope(env->isolate());
+  Context::Scope context_scope(env->context());
+
+  int status = SSL_select_next_proto(const_cast<unsigned char**>(out), outlen,
+                                     w->alpn_protos_, w->alpn_protos_len_,
+                                     in, inlen);
+
+  switch (status) {
+    case OPENSSL_NPN_NO_OVERLAP:
+      // According to 3.2. Protocol Selection of RFC7301,
+      // fatal no_application_protocol alert shall be sent
+      // but current openssl does not support it yet. See
+      // https://rt.openssl.org/Ticket/Display.html?id=3463&user=guest&pass=guest
+      // Instead, we send a warning alert for now.
+      return SSL_TLSEXT_ERR_ALERT_WARNING;
+    case OPENSSL_NPN_NEGOTIATED:
+      return SSL_TLSEXT_ERR_OK;
+    default:
+      return SSL_TLSEXT_ERR_ALERT_FATAL;
+  }
+}
+#endif  // TLSEXT_TYPE_application_layer_protocol_negotiation
+
+
+template <class Base>
+void SSLWrap<Base>::GetALPNNegotiatedProto(
+    const FunctionCallbackInfo<v8::Value>& args) {
+#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
+  HandleScope scope(args.GetIsolate());
+  Base* w = Unwrap<Base>(args.Holder());
+
+  const unsigned char* alpn_proto;
+  unsigned int alpn_proto_len;
+
+  SSL_get0_alpn_selected(w->ssl_, &alpn_proto, &alpn_proto_len);
+
+  if (!alpn_proto)
+    return args.GetReturnValue().Set(false);
+
+  args.GetReturnValue().Set(
+      OneByteString(args.GetIsolate(), alpn_proto, alpn_proto_len));
+#endif  // TLSEXT_TYPE_application_layer_protocol_negotiation
+}
+
+
+template <class Base>
+void SSLWrap<Base>::SetALPNProtocols(
+    const FunctionCallbackInfo<v8::Value>& args) {
+#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
+  HandleScope scope(args.GetIsolate());
+  Base* w = Unwrap<Base>(args.Holder());
+  if (args.Length() < 1 || !Buffer::HasInstance(args[0]))
+    return w->env()->ThrowTypeError("Must give a Buffer as first argument");
+
+  if (w->is_client()) {
+    const unsigned char* alpn_protos =
+        reinterpret_cast<const unsigned char*>(Buffer::Data(args[0]));
+    unsigned alpn_protos_len = Buffer::Length(args[0]);
+    int r = SSL_set_alpn_protos(w->ssl_, alpn_protos, alpn_protos_len);
+    CHECK_EQ(r, 0);
+  } else {
+    w->alpn_protos_ =
+        reinterpret_cast<const unsigned char*>(Buffer::Data(args[0]));
```

Comment:
It seems that `JSObject::SetHiddenProperty` is not a public API so is it done by using `Object.defineProperty` in JS land like following diff? Or is it something else?
```diff
diff --git a/lib/_tls_legacy.js b/lib/_tls_legacy.js
index 8c079e3..b3f6385 100644
--- a/lib/_tls_legacy.js
+++ b/lib/_tls_legacy.js
@@ -727,8 +727,14 @@ function SecurePair(context, isServer, requestCert, rejectUnauthorized,
   }

   if (process.features.tls_alpn && options.ALPNProtocols) {
-    // keep reference in secureContext not to be GC-ed
-    this.ssl._secureContext.alpnBuffer = options.ALPNProtocols;
+    // Keep reference in a hidden property of secureContext not to be
+    // modified by user and GC-ed.
+    Object.defineProperty(this.ssl._secureContext, 'alpnBuffer', {
+      enumerable: false,
+      configurable: false,
+      writable: false,
+      value: options.ALPNProtocols
+    });
     this.ssl.setALPNrotocols(this.ssl._secureContext.alpnBuffer);
     this.alpnProtocol = null;
   }
diff --git a/lib/_tls_wrap.js b/lib/_tls_wrap.js
index d918656..f2438cf 100644
--- a/lib/_tls_wrap.js
+++ b/lib/_tls_wrap.js
@@ -455,8 +455,14 @@ TLSSocket.prototype._init = function(socket, wrap) {
     ssl.setNPNProtocols(options.NPNProtocols);

   if (process.features.tls_alpn && options.ALPNProtocols) {
-    // keep reference in secureContext not to be GC-ed
-    ssl._secureContext.alpnBuffer = options.ALPNProtocols;
+    // Keep reference in a hidden property of secureContext not to be
+    // modified by user and GC-ed.
+    Object.defineProperty(ssl._secureContext, 'alpnBuffer', {
+      enumerable: false,
+      configurable: false,
+      writable: false,
+      value: options.ALPNProtocols
+    });
     ssl.setALPNProtocols(ssl._secureContext.alpnBuffer);
   }
````